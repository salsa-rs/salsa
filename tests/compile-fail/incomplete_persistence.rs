#[salsa::tracked(persist)]
struct Persistable<'db> {
    field: NotPersistable<'db>,
}

#[salsa::tracked]
struct NotPersistable<'db> {
    field: usize,
}

#[salsa::tracked(persist)]
fn query(_db: &dyn salsa::Database, _input: NotPersistable<'_>) {}

fn main() {}
