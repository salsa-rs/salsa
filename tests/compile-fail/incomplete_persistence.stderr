error[E0277]: the trait bound `NotPersistable<'_>: Serialize` is not satisfied
 --> tests/compile-fail/incomplete_persistence.rs:1:1
  |
1 | #[salsa::tracked(persist)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^
  | |
  | the trait `Serialize` is not implemented for `NotPersistable<'_>`
  | required by a bound introduced by this call
  |
  = note: for local types consider adding `#[derive(serde::Serialize)]` to your `NotPersistable<'_>` type
  = note: for types from other crates check whether the crate offers a `serde` feature flag
  = help: the following other types implement trait `Serialize`:
            &'a T
            &'a mut T
            ()
            (T,)
            (T0, T1)
            (T0, T1, T2)
            (T0, T1, T2, T3)
            (T0, T1, T2, T3, T4)
          and $N others
  = note: required for `(NotPersistable<'_>,)` to implement `Serialize`
  = note: this error originates in the macro `salsa::plumbing::setup_tracked_struct` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NotPersistable<'_>: Deserialize<'_>` is not satisfied
 --> tests/compile-fail/incomplete_persistence.rs:1:1
  |
1 | #[salsa::tracked(persist)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Deserialize<'_>` is not implemented for `NotPersistable<'_>`
  |
  = note: for local types consider adding `#[derive(serde::Deserialize)]` to your `NotPersistable<'_>` type
  = note: for types from other crates check whether the crate offers a `serde` feature flag
  = help: the following other types implement trait `Deserialize<'de>`:
            &'a Path
            &'a [u8]
            &'a str
            ()
            (T,)
            (T0, T1)
            (T0, T1, T2)
            (T0, T1, T2, T3)
          and $N others
  = note: required for `(NotPersistable<'_>,)` to implement `Deserialize<'_>`
  = note: this error originates in the macro `salsa::plumbing::setup_tracked_struct` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `NotPersistable<'db>: Serialize` is not satisfied
  --> tests/compile-fail/incomplete_persistence.rs:12:45
   |
12 | fn query(_db: &dyn salsa::Database, _input: NotPersistable<'_>) {}
   |                                             ^^^^^^^^^^^^^^^^^^ the trait `Serialize` is not implemented for `NotPersistable<'db>`
   |
   = note: for local types consider adding `#[derive(serde::Serialize)]` to your `NotPersistable<'db>` type
   = note: for types from other crates check whether the crate offers a `serde` feature flag
   = help: the following other types implement trait `Serialize`:
             &'a T
             &'a mut T
             ()
             (T,)
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
             (T0, T1, T2, T3, T4)
           and $N others
note: required by a bound in `query_input_is_persistable`
  --> tests/compile-fail/incomplete_persistence.rs:11:1
   |
11 | #[salsa::tracked(persist)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | required by a bound in this function
   | required by this bound in `query_input_is_persistable`
   = note: this error originates in the macro `salsa::plumbing::setup_tracked_fn` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `for<'de> NotPersistable<'db>: Deserialize<'de>` is not satisfied
  --> tests/compile-fail/incomplete_persistence.rs:12:45
   |
12 | fn query(_db: &dyn salsa::Database, _input: NotPersistable<'_>) {}
   |                                             ^^^^^^^^^^^^^^^^^^ the trait `for<'de> Deserialize<'de>` is not implemented for `NotPersistable<'db>`
   |
   = note: for local types consider adding `#[derive(serde::Deserialize)]` to your `NotPersistable<'db>` type
   = note: for types from other crates check whether the crate offers a `serde` feature flag
   = help: the following other types implement trait `Deserialize<'de>`:
             &'a Path
             &'a [u8]
             &'a str
             ()
             (T,)
             (T0, T1)
             (T0, T1, T2)
             (T0, T1, T2, T3)
           and $N others
note: required by a bound in `query_input_is_persistable`
  --> tests/compile-fail/incomplete_persistence.rs:11:1
   |
11 | #[salsa::tracked(persist)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^
   | |
   | required by a bound in this function
   | required by this bound in `query_input_is_persistable`
   = note: this error originates in the macro `salsa::plumbing::setup_tracked_fn` which comes from the expansion of the attribute macro `salsa::tracked` (in Nightly builds, run with -Z macro-backtrace for more info)
